File: generated_code/GeminiPro/Java/041_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the property that
 * recently accessed elements are quick to access again. It performs basic
 * operations such as insertion, deletion, and search. After a search, insertion,
 * or deletion, the accessed node is moved to the root of the tree through a
 * series of rotations in an operation called "splaying".
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * This is a private inner class, making the SplayTree class self-contained.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructor for a Node.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Core Splay Operations ---

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. It brings the node with the given key
     * (or the last accessed node if the key is not found) to the root of the tree.
     * This is a recursive top-down splaying implementation.
     *
     * @param node The root of the current subtree.
     * @param key The key to splay towards the root.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        // Base case: node is null or key is at the root
        if (node == null || node.key == key) {
            return node;
        }

        // Key is in the left subtree
        if (key < node.key) {
            // Key is not in the tree, we are done
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left-Left)
            if (key < node.left.key) {
                // Recursively bring the key as root of left-left
                node.left.left = splay(node.left.left, key);
                // First rotation for root
                node = rightRotate(node);
            }
            // Zig-Zag (Left-Right)
            else if (key > node.left.key) {
                // Recursively bring the key as root of left-right
                node.left.right = splay(node.left.right, key);
                // First rotation for node.left
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }

            // Do second rotation for root (if possible)
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key is in the right subtree
        else { // key > node.key
            // Key is not in the tree, we are done
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right-Left)
            if (key < node.right.key) {
                // Recursively bring the key as root of right-left
                node.right.left = splay(node.right.left, key);
                // First rotation for node.right
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > node.right.key) {
                // Recursively bring the key as root of right-right
                node.right.right = splay(node.right.right, key);
                // First rotation for root
                node = leftRotate(node);
            }

            // Do second rotation for root (if possible)
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    // --- Public API Methods ---

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree is splayed on that key but otherwise remains unchanged.
     * After insertion of a new key, the new node becomes the root of the tree.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the closest node to the root.
        root = splay(root, key);

        // If key is already present, do nothing more.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not found, the tree is splayed on the last accessed node
     * but otherwise remains unchanged.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        // Splay the tree for the key. If found, it becomes the root.
        root = splay(root, key);

        // If key is not in the tree (or tree is empty), do nothing.
        if (root == null || root.key != key) {
            return;
        }

        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Find the max element in the left subtree by splaying for the largest possible key (our deleted key)
            // This brings the predecessor of the deleted key to the root of the left subtree.
            leftSubtree = splay(leftSubtree, key);
            // After splaying, the new root of the left subtree is the max element and has no right child.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }

    /**
     * Searches for a key in the tree.
     * This operation performs a splay on the accessed node (or the last node on the search path),
     * moving it to the root of the tree.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return key;
        }
        return null;
    }
}

File: generated_code/GeminiPro/Python/041_SplayTree.py
import sys

# It's good practice to increase recursion limit for tree structures,
# though this implementation is iterative to avoid it.
# sys.setrecursionlimit(10**6)

class SplayTree:
    """
    A self-contained Python class that implements a Splay Tree.
    
    This class provides a dictionary-like set for storing unique integer keys.
    It includes insert, delete, and search operations. The key feature is the
    splaying operation: whenever a key is accessed (searched, inserted, or
    deleted), the corresponding node is moved to the root of the tree to
    optimize future accesses for that key.
    """

    class _Node:
        """A private inner class representing a node in the Splay Tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x
        
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splaying operation on node x, moving it to the root.
        This is the core operation that gives the Splay Tree its self-balancing
        and self-optimizing properties.
        """
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:  # Zig case (parent is root)
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig case (left-left)
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig case (right-right)
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag case (left-right)
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag case (right-left)
                self._right_rotate(p)
                self._left_rotate(g)

    def search(self, key):
        """
        Searches for a key in the tree.
        
        Performs the splaying operation on the accessed node if found. If the
        node is not found, its would-be parent (the last node visited) is
        splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # Key not found, splay the last visited node (the would-be parent)
        if last_node:
            self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted as in a standard BST, and then the
        new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay it and we're done
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # Create the new node
        new_node = self._Node(key, parent)

        # Insert the new node into the tree
        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.
        
        First, it searches for the key, which splays the node to the root if
        found. If the key is not found, its parent is splayed and the operation
        ends. If the key is found and splayed, it is removed, and its two
        subtrees are merged.

        Args:
            key: The integer key to delete.
        """
        # Search for the key. This also splays the node (or its parent) to the root.
        if not self.search(key):
            # Key was not found. search() already splayed the would-be parent.
            # There's nothing to delete.
            return

        # After a successful search(), the node with the key is at the root.
        node_to_delete = self.root
        
        left_subtree = node_to_delete.left
        right_subtree = node_to_delete.right

        if not left_subtree:
            # No left child, so the right subtree becomes the new tree.
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect the left subtree from the old root.
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree.
            max_node_in_left = left_subtree
            while max_node_in_left.right:
                max_node_in_left = max_node_in_left.right
            
            # Splay this maximum node to the root of the (temporary) left subtree.
            # This is a common technique: temporarily treat the subtree as the
            # whole tree to reuse the _splay method.
            self.root = left_subtree
            self._splay(max_node_in_left)
            
            # After splaying, self.root is the new root of the merged tree.
            # This new root (max_node_in_left) has no right child.
            # Attach the original right subtree as its right child.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root